core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreMessageHandler.java

```diff
@@ -122,46 +124,26 @@ private void handleClearSubject(ClearSubjectValue clearSubjectValue) {
     }
   }
 
-  private void handleSchemaUpdate(SchemaKey schemaKey, SchemaValue schemaObj) {
-    if (schemaObj != null) {
+  private void handleSchemaUpdate(SchemaKey schemaKey,
+                                  SchemaValue schemaValue,
+                                  SchemaValue oldSchemaValue) {
+    if (schemaValue != null) {
       // If the schema is marked to be deleted, we store it in an internal datastructure
       // that holds all deleted schema keys for an id.
       // Whenever we encounter a new schema for a subject, we check to see if the same schema
       // (same id) was deleted for the subject ever. If so, we tombstone those delete keys.
       // This helps optimize the storage. The main reason we only allow soft deletes in SR is that
       // consumers should be able to access the schemas by id. This is guaranteed when the schema is
       // re-registered again and hence we can tombstone the record.
-      if (schemaObj.isDeleted()) {
-        this.lookupCache.schemaDeleted(schemaKey, schemaObj);
+      if (schemaValue.isDeleted()) {
+        this.lookupCache.schemaDeleted(schemaKey, schemaValue);
       } else {
         // Update the maximum id seen so far
-        idGenerator.schemaRegistered(schemaKey, schemaObj);
-        lookupCache.schemaRegistered(schemaKey, schemaObj);
-        List<SchemaKey> schemaKeys = lookupCache.deletedSchemaKeys(schemaObj);
-        schemaKeys.stream().filter(v ->
-            v.getSubject().equals(schemaObj.getSubject())
-                && v.getVersion() != schemaObj.getVersion())
-            .forEach(this::tombstoneSchemaKey);
+        idGenerator.schemaRegistered(schemaKey, schemaValue);
+        lookupCache.schemaRegistered(schemaKey, schemaValue);
       }
     } else {
-      lookupCache.schemaTombstoned(schemaKey);
-    }
-  }
-
-  private void tombstoneSchemaKey(SchemaKey schemaKey) {
-    if (schemaRegistry.getKafkaStore().initialized()) {
-      tombstoneExecutor.execute(() -> {
-            try {
-              schemaRegistry.getKafkaStore().waitForInit();
-              schemaRegistry.getKafkaStore().delete(schemaKey);
-              log.debug("Tombstoned {}", schemaKey);
-            } catch (InterruptedException e) {
-              log.error("Interrupted while waiting for the tombstone thread to be initialized ", e);
-            } catch (StoreException e) {
-              log.error("Failed to tombstone {}", schemaKey, e);
-            }
-          }
-      );
```

Comment:
This is the old method of tombstoning, which has moved to `register`.  This code actually had a bug where both masters and replicas would perform tombstoning, leading to redundant tombstones.  By moving the code to `register`, this problem is avoided.